### PR TITLE
Feature/lab1d

### PR DESCRIPTION
**Pull request by decel337**

### CODE REVIEW

File: RenderEngine/RenderEngine/Core/Renderer.cs [hunk 1]
```diff
@@ -51,6 +51,8 @@ public Renderer(Camera camera, Scene scene)
                             intersectionPoint = intersection;
                             minSquareDistance = squareDist;
                         }
+
+                        screen[i, j] = Scene.Lighting.FirstOrDefault().GetLight(shape, intersectionPoint);
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Core/Renderer.cs:55]: FirstOrDefault returns nullable<T>, possible null reference exception

File: RenderEngine/RenderEngine/Shapes/Disk.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                if (t >= 0)
+                    isPlaneIntersect = true;
+            }
+            if (!isPlaneIntersect)
+                return null;
+            // check intersect with disk
+            Vector3 vectorIntersect = ray.GetPoint(t);
+            Vector3 distance = vectorIntersect - Orig;
+            float distanceValue = Vector3.Dot(distance, distance);
+            if(distanceValue > Radius * Radius)
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Disk.cs:43]: use conditional expression

File: RenderEngine/RenderEngine/Shapes/Sphere.cs [hunk 3]
```diff
@@ -45,5 +45,15 @@ public Sphere(Vector3 orig, float radius)
 
             return t2 > 0 ? ray.GetPoint(t2) : null;
         }
+
+        public Vector3 GetNormal(Vector3? intersectionPoint)
+        {
+            if(intersectionPoint == null)
+            {
+                return new Vector3(0, 0, 0);
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Sphere.cs:53]: Vector3 has a static Zero field

File: RenderEngine/RenderEngine/Shapes/Triangle.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RenderEngine.Shapes
+{
+    public record Triangle : IShape
+    {
+        public Vector3 v0 { get; init; }
+        public Vector3 v1 { get; init; }
+        public Vector3 v2 { get; init; }
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Triangle.cs:15]: public properties - CamelCase

File: RenderEngine/RenderEngine/Shapes/Triangle.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        public Triangle(Vector3 v0, Vector3 v1, Vector3 v2)
+        {
+            this.v0 = v0;
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Triangle.cs:19]: don't use the 'this' keyword

File: RenderEngine/RenderEngine/Shapes/Triangle.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            var v = Vector3.Dot(ray.Dir, qvec) * invDet;
+
+            if (v < 0 || u + v > 1)
+            {
+                return null;
+            }
+
+            var t = Vector3.Dot(edge2, qvec) * invDet;
+
+            if(t <= tolerance)
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Triangle.cs:61]: also can be replaced with conditional expression

File: RenderEngine/RenderEngine/Shapes/Triangle.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            return ray.GetPoint(t);
+        }
+
+        public Vector3 GetNormal(Vector3? intersectionPoint)
+        {
+            var edge1 = v1 - v0;
+            var edge2 = v2 - v0;
+
+            var pvec = Vector3.Cross(edge1, edge2);
+            return pvec;
```
Comment (st1lson) [inline RenderEngine/RenderEngine/Shapes/Triangle.cs:75]:
```suggestion
            return Vector3.Cross(edge1, edge2);
```

**Review by st1lson — CHANGES_REQUESTED** [review]

File: RenderEngine/RenderEngine/Basic/Point3.cs [hunk 8]
```diff
@@ -22,6 +22,11 @@ public Point3(float x, float y, float z)
         return new Point3(left.X + right.X, left.Y + right.Y, left.Z + right.Z);
     }
 
+    public static Point3 operator -(Point3 left, Point3 right)
```
Comment (Akara4ok) [inline RenderEngine/RenderEngine/Basic/Point3.cs:25]: Invalid operator for point

File: RenderEngine/RenderEngine/Core/Renderer.cs (hunk 1 above)
Comment (Akara4ok) [inline RenderEngine/RenderEngine/Core/Renderer.cs:55]: You have to remember the closest shape and find color for it only in the end

File: RenderEngine/RenderEngine/Lightings/Light.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+using RenderEngine.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RenderEngine.Lightings
+{
+    public class Light : ILighting
```
Comment (Akara4ok) [inline RenderEngine/RenderEngine/Lightings/Light.cs:11]: Rename to directional light

File: RenderEngine/RenderEngine.Console/Program.cs [hunk 10]
```diff
@@ -16,10 +16,14 @@ private static void Main(string[] args)
...
             );
 
-        List<IShape> shapes = new List<IShape>() { new Sphere(new Vector3(-3, 5, -3), 3) };
+        List<IShape> shapes = new List<IShape>() {
+            new Sphere(new Vector3(-3, 5, -3), 3),
+            new Disk(new Vector3(3, 5, 3), 3, new Vector3(0, -1, 0)),
+            new Triangle(new Vector3(0, 5, 0), new Vector3(1, 5, 0), new Vector3(0, 5, 1))
+        };
 
         //TODO: add some lighting
```
Comment (Akara4ok) [inline RenderEngine/RenderEngine.Console/Program.cs:25]: Delete this comments

File: RenderEngine/RenderEngine/Shapes/Sphere.cs [hunk 11]
```diff
@@ -45,5 +45,15 @@ public Sphere(Vector3 orig, float radius)
 
             return t2 > 0 ? ray.GetPoint(t2) : null;
         }
+
+        public Vector3 GetNormal(Vector3? intersectionPoint)
```
Comment (Akara4ok) [inline RenderEngine/RenderEngine/Shapes/Sphere.cs:49]: delete optional value here but use Vector.Value when you pass argument

**Review by st1lson — APPROVED** [review]

**Review by Akara4ok — APPROVED** [review]

**Pull request merged**